### PR TITLE
JDK-8299470: sun/jvm/hotspot/SALauncher.java handling of negative rmiport args

### PR DESCRIPTION
**Pull request by MBaesken**
The test serviceability/sa/sadebugd/SADebugDTest.java can pass under some circumstances a negative rmiport (--rmiport -1) to SALauncher.java.
This leads to a somewhat misleading message 
`[debugd] Argument is expected for 'rmiport' `
(we set an argument [-1] but probably this is not what is really expected) and additionally the real exception is not shown.
Probably also a warning in case of negative rmiport values should be printed because they seem to lead to errors.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299470](https://bugs.openjdk.org/browse/JDK-8299470): sun/jvm/hotspot/SALauncher.java handling of negative rmiport args

### Reviewers
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**) ⚠️ Review applies to [d42cc159](https://git.openjdk.org/jdk/pull/11811/files/d42cc1596afe3b8fd9d1625057d9f38f09be10c7)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11811/head:pull/11811` \
`$ git checkout pull/11811`

Update a local copy of the PR: \
`$ git checkout pull/11811` \
`$ git pull https://git.openjdk.org/jdk pull/11811/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11811`

View PR using the GUI difftool: \
`$ git pr show -t 11811`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11811.diff">https://git.openjdk.org/jdk/pull/11811.diff</a>

</details>

### CODE REVIEW

**Comment by MBaesken** [comment]
I adjusted the exception message a bit to show what is wrong (-1 as an argument is not expected/supported).
Removed one warning output because it seems we do not get there.

Any reviews ?

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java [hunk 1]
```diff
@@ -500,6 +500,8 @@ public static void main(String[] args) {
                 func.accept(oldArgs);
             }
         } catch (SAGetoptException e) {
+            System.err.println("SA agent option related exception occured");
```
Comment (RealCLanger) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:503]: Maybe you could do here:
`System.err.println("SA agent option exception occured: " + e.getMessage);`
and remove the `System.err.println(e.getMessage());` in line 505.

**Review by RealCLanger — CHANGES_REQUESTED** [review]
I've got one suggestion, plus, the copyright year now is 2023 😉

**Review by RealCLanger — APPROVED** [review]

**Comment by kevinjwalls** [comment]
The changes here to log a better error are I think good (maybe adding the stacktrace printing is not needed now the error is more understandable!), but aborting the launch attempt in the caller, if it gets -1 when trying to find a free port, should be even clearer?

In
test/hotspot/jtreg/serviceability/sa/sadebugd/SADebugDTest.java

the method testWithPid() finds an rmiPort:

  final int rmiPort = useRmiPort ? Utils.findUnreservedFreePort(REGISTRY_DEFAULT_PORT, registryPort) : -1;

...but can proceed to launch even if it gets that -1 value?   That looks like a mistake.

In the "if (useRmiPort) {" block, we should be failing the test if rmiPort is -1, saying something like
"cannot find an rmiPort, findUnreservedFreePort returns -1"

A similar abort if setting registryPort gets -1 might also be good?

**Comment by MBaesken** [comment]
> In the "if (useRmiPort) {" block, we should be failing the test if rmiPort is -1, saying something like "cannot find an rmiPort, findUnreservedFreePort returns -1"
> 
> A similar abort if setting registryPort gets -1 might also be good?

Hi Kevin,  I think you are correct, adjusting the  test test/hotspot/jtreg/serviceability/sa/sadebugd/SADebugDTest.java  in those two cases (rmiPort / registryPort  -1)  probably makes sense.  Should we do it in another JBS issue or here ?
Additionally , should we fail the test in these cases , or pass with some warning - message ?

**Comment by kevinjwalls** [comment]
SADebugDTest is only one test, so seems OK to have it fail as soon as we realise we need a port, and it has a value of -1.

I would do it in this change as they are so connected, but really whichever works best for you.  (I don't see other uses of findUnreservedFreePort() in the same test hierarchy, so this task should not keep on growing... 8-) )

**Comment by kevinjwalls** [comment]
Looks good!

Rereading the change, occured should be "occurred" (we do have a lot of "occured" in the JDK, but many more "occurred")
On the line after that, I don't think we should introduce the stacktrace.  The user is getting a more readable error message already...
In SAGetopt.java you single-quote the 'opt' value, but could we do the same to the  _argv[_optind] value.

Then can you could make it "a registryPort" rather than "an", and I think I'm done with comments.

Thanks!

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

**Comment by kevinjwalls** [comment]
Thanks for updating. 
I do think we should remove the e.printStackTrace() that was added in src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java
as this is a message for the user.
In the test, those new exceptions will produce a stacktrace for when we look at the failing test.

**Review by kevinjwalls — APPROVED** [review]

**Comment by dholmes-ora** [comment]
This change has introduced a test failure in our tier 5 CI testing:

https://bugs.openjdk.org/browse/JDK-8299700

**Comment by dholmes-ora** [comment]
I missed an issue had already been filed for the failure.

**Pull request closed without merging**